core/src/test/scala/unit/kafka/server/LeaderElectionTest.scala

```diff
@@ -96,7 +96,28 @@ class LeaderElectionTest extends ZooKeeperTestHarness {
       assertEquals("Second epoch value should be %d".format(leaderEpoch1+1) , leaderEpoch1+1, leaderEpoch2)
 
     servers.last.startup()
+    //make sure second server joins the ISR
+    TestUtils.waitUntilTrue(() => {
+      val partitionInfoOpt = servers.last.metadataCache.getPartitionInfo(topic, partitionId)
+      if (partitionInfoOpt.isDefined) {
+        partitionInfoOpt.get.basePartitionState.isr.size() == 2
+      } else {
+        false
+      }
+    }, "Inconsistent metadata after second broker startup")
+
     servers.head.shutdown()
+    //make sure first server is removed from the ISR
+    TestUtils.waitUntilTrue(() => {
```

Comment:
Thinking about this a bit more. I am not sure this waitUntilTrue() is useful. We can't guarantee whether the controller, if on first server can act on the broker failure event before the first server is shutdown. We just need to make sure that PartitionStateMachine and ReplicaStateMachine transition the partition and the replica properly either by the old or the new controller, which is already done by the assertion in line 132.

A few other comments about the existing test, not introduced by this PR.

line 80. It seems that a better way to avoid transient test failure is to make sure that both brokers are registered in the controller before creating the topic. Then, we know for sure that the preferred replica is the initial leader, which can simplify the rest of the test.

line 84. The comment seems inaccurate since the last broker is not the preferred replica

Reply:
@junrao Thanks for the review. I have cleanup the testLeaderElectionAndEpoch test. Pls take one more look.